File: Range Sum of BST_bfs.py
#time: O(n)
#space: O(m) where m is the no.of leaf nodes

#bfs


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rangeSumBST(self, root: TreeNode, low: int, high: int) -> int:
        q=deque()
        sum1=0
        q.append(root)
        while(q!=deque()):
            size=len(q)
            for i in range(size):
                r=q.popleft()
                if(low<=r.val<=high):
                    sum1+=r.val
                if(r.left!=None):
                    q.append(r.left)
                if(r.right!=None):
                    q.append(r.right)
        return sum1
            
            

File: RangeSumOfBST/RangeSumOfBST_BFS.java
/* Time Complexity : O(n) 
*   n - total number of elements in a tree */
/* Space Complexity : O(h) 
*   h - size of the queue */
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

//BFS
class Solution {
    public int rangeSumBST(TreeNode root, int low, int high) {
        int res = 0;
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        while(!q.isEmpty()){
            TreeNode curr = q.poll();
            if(curr.val >= low && curr.val <= high){
                res += curr.val;
            }
            if(curr.left != null && curr.val >= low){
                q.add(curr.left);
            }
            if(curr.right != null && curr.val <= high){
                q.add(curr.right);
            }
        }
        return res;
    }
}